deviceplugin: close grpc conn at stop

dqminh opened the pull request:
Also remove useless loop in device_plugin_stub, it doesnt do anything
except looping for full 10s in most case. It fixes some errors but i
think its because of connection not closed properly. Removing the loop
also cut test runtime from 60s to 6s on my machine.

With this `go test -count 20 -race` run successfully on my machine
without any errors.

**Release note**:
```release-note
NONE
```

feiskyer commented on pkg/kubelet/deviceplugin/device_plugin_stub.go:
```diff
@@ -67,14 +67,6 @@ func (m *Stub) Start() error {
 	go m.server.Serve(sock)
-	// Wait till grpc server is ready.
-	for i := 0; i < 10; i++ {
-		services := m.server.GetServiceInfo()
```
Will removing this cause problems? e.g. if server is not ready, then some calls on it may fail.

dqminh replied:
indeed. But doing this is incorrect as it just wait for 10s all the time. GRPC has a `FailFast` toggle that we can set to false to allow the connection to be retried. Usually the toggle is set to not fail fast as part of health check process ( that we can also use in test ) to determine if the server is responding within a time period.
But the current API doesn't expose the healthcheck API that we can set `grpc.FailFast(false)` on. Not sure if we should add one as part if this change.

dqminh committed:
deviceplugin: new sockets for each test

Reusing the same socket for grpc calls in tests can lead to transient
failures when attempting calls on an unlinking socket (no such file).

```
grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /tmp/mock.sock: connect: no such file or directory"; Reconnecting to {/tmp/mock.sock <nil>}
E0925 11:58:41.469339     589 endpoint.go:89] listAndWatch ended unexpectedly for device plugin mock with error rpc error: code = Unavailable desc = grpc: the connection is unavailable
--- FAIL: TestList (0.00s)
        Error Trace:    endpoint_test.go:55
        Error:          Received unexpected error:
                        rpc error: code = Unavailable desc = grpc: the connection is unavailable
```

Instead generate a new name for each test, this seems to remove the odd
random errors in tests.
Additionally we should also check for server availability in tests, but
the current RPC scheme doesnt exactly have a health check endpoint that
is usually used for such purpose (together with grpc.FailFast).

Also makes registraion tests failed within timeout instead of waiting
for the channel forever.

dqminh commented:
/retest

ncdc commented:
/unassign
/assign @derekwaynecarr

RenaudWasTaken commented:
Hey @dqminh thanks for your contribution!

Sorry but closing GRPC connexion at stop was already done in another [PR](https://github.com/kubernetes/kubernetes/pull/52747)

On the other hand, I like the test improvements you made! 
@jiayingz for more thoughts on the other additions of this PR.

dqminh commented:
@RenaudWasTaken many apologies for neglecting this change. It completely fell off my mind. I will see if i can rebase only the test changes.

dqminh commented:
@RenaudWasTaken @jiayingz this was rebased with only the test changes. I have also removed changes from https://github.com/kubernetes/kubernetes/pull/53028 since the changes to not reuse unix socket endpoint should likely fix all the races we have here instead of retrying more.

Also seeing some relevant changes in https://github.com/kubernetes/kubernetes/pull/52616. I havent looked into it closely but @RenaudWasTaken feel free to close this and move the related changes together if you want to keep them together.

jiayingz commented on pkg/kubelet/cm/deviceplugin/manager_test.go:
```diff
@@ -66,38 +87,41 @@ func TestDevicePluginReRegistration(t *testing.T) {
+	select {
+	case <-callbackChan:
+	case <-time.After(5 * time.Second):
+		t.FailNow()
```
cc @lichuqiang 
There is actually a race here that we can't just wait for callback before checking m.Devices(). Basically, with the current code, we can get a callback from endpoint.list() before the endpoint is added to the manager. With PR https://github.com/kubernetes/kubernetes/pull/52149 this race is gone so we can remove this wait. I would suggest we take this part of change out of this PR, wait after https://github.com/kubernetes/kubernetes/pull/52149 gets in, and then clean up this test.

feiskyer commented:
/assign @jiayingz

jiayingz commented:
@dqminh I think most part of this PR has been merged in with other PRs. Do you think if this PR can be closed now?

dqminh commented:
@jiayingz yes, i think its better to close this one now. I will double check and open a new one if necessary. Thanks for the notice !
